### PR TITLE
Restructure estimator tests

### PR DESCRIPTION
**Pull request by AlexanderBenz**
See #303 

**Additional context**
Include tests for classifier and regression

### CODE REVIEW

File: skactiveml/tests/template_estimator.py [hunk 1]
```diff
@@ -0,0 +1,686 @@
...
+            ([1.0, 1.1, 0.9], None),
+        ]
+        super().test_fit_param_y(test_cases)
+        # TODO: Test schould throw a TypeError
+        # Wrapper classes are failing because of
+        # "numpy.core._exceptions._UFuncNoLoopError: ufunc 'add' did not
+        # contain a loop with signature matching types
+        # (dtype('<U32'), dtype('<U32')) -> None"
+
+        # test_cases = [([1.0, "nan", 0.9], None)]
```
Comment (AlexanderBenz) [inline skactiveml/tests/template_estimator.py:574]: @mherde

File: skactiveml/regressor/tests/test_nic_kernel_regressor.py [hunk 2]
```diff
@@ -9,14 +9,120 @@
...
+        y = norm.rvs(loc=1.24, scale=0.0245, size=500, random_state=0)
+        if self.estimator_class_string == "NICKernelRegressor":
+            reg = self.estimator_class(kappa_0=0, nu_0=0)
+            reg.fit(X, y)
+            mu, sigma = reg.predict([[0]], return_std=True)
+            np.testing.assert_almost_equal(mu, 1.24, decimal=3)
+            np.testing.assert_almost_equal(sigma, 0.0245, decimal=3)
+
+    def test_predict_target_distribution(self):
+        # TODO: Test is missing
```
Comment (AlexanderBenz) [inline skactiveml/regressor/tests/test_nic_kernel_regressor.py:107]: @mherde

**Review by tpham93 — CHANGES_REQUESTED** [review]

File: skactiveml/tests/template_estimator.py [hunk 3]
```diff
@@ -0,0 +1,686 @@
...
+        # Check if predict_proba is being tested.
+        with self.subTest(msg="test_predict_freq"):
+            self.assertTrue(
+                hasattr(self, "test_predict_freq"),
+                msg=f"'test_predict_freq' missing in {self.__class__}",
+            )
+        super().test_param_test_availability()
+
+    def test_predict_freq_param_X(self, test_cases=None):
+        if hasattr(self.estimator_class, "predict_freq"):
```
Comment (tpham93) [inline skactiveml/tests/template_estimator.py:520]:
```suggestion
```

**Review by tpham93 — CHANGES_REQUESTED** [review]

File: skactiveml/regressor/tests/test_nic_kernel_regressor.py [hunk 4]
```diff
@@ -9,14 +9,120 @@
...
+
+class TestNICKernelEstimator(
+    TemplateTestNICKernelEstimator, unittest.TestCase
+):
     def setUp(self):
-        self.X = np.array([[0, 1], [1, 0], [2, 3]])
-        self.y = np.array([1, 2, 3])
-        self.X_cand = np.array([[2, 1], [3, 5]])
+        estimator_class = NICKernelRegressor
+        self.estimator_class_string = "NICKernelRegressor"
```
Comment (mherde) [inline skactiveml/regressor/tests/test_nic_kernel_regressor.py:116]: What is the purpose of this string?

File: skactiveml/regressor/tests/test_nic_kernel_regressor.py [hunk 5]
```diff
@@ -75,63 +184,33 @@ def test_random_state(self):
...
-        np.testing.assert_almost_equal(sigma, 0.0245, decimal=3)
-
-
-class TestNadarayaWatsonRegressor(unittest.TestCase):
+class TestNadarayaWatsonRegressor(
+    TemplateTestNICKernelEstimator, unittest.TestCase
+):
     def setUp(self):
+        estimator_class = NadarayaWatsonRegressor
+        self.estimator_class_string = "NadarayaWatsonRegressor"
```
Comment (mherde) [inline skactiveml/regressor/tests/test_nic_kernel_regressor.py:193]: Do we need this string?

File: skactiveml/regressor/tests/test_wrapper.py [hunk 6]
```diff
@@ -191,3 +255,38 @@ def test_predict_target_distribution(self):
         self.assertRaises(
             ValueError, reg.predict_target_distribution, self.X_cand
         )
+
+    def test_fit(self):
```
Comment (mherde) [inline skactiveml/regressor/tests/test_wrapper.py:259]: Whats is the purpose of this test?

**Review by mherde — CHANGES_REQUESTED** [review]

File: skactiveml/classifier/tests/test_wrapper.py [hunk 7]
```diff
@@ -199,50 +217,263 @@ def test_pipeline(self):
...
-            estimator=ParzenWindowClassifier(), ignore_estimator_partial_fit=0
+
+    def test_fit_param_X(self, test_cases=None, replace_init_params=None):
+        replace_init_params = {
+            "classes": [0, 1],
+            "missing_label": -1,
+            "estimator": SklearnClassifier(
+                MixtureModelClassifier(), missing_label=-1, classes=[0, 1]
+            ),
+        }
```
[the remark this answers is not in the document]
Comment (tpham93) [reply skactiveml/classifier/tests/test_wrapper.py:387]: @AlexanderBenz Can you have a look at this? We don't understand why we need to override the init parameters for test_fit_param_X

File: skactiveml/tests/template_estimator.py [hunk 8]
```diff
@@ -247,8 +248,7 @@ def _test_param(
                     init_params[key] = val
```
Comment (mherde) [inline skactiveml/tests/template_estimator.py:248]: @AlexanderBenz 
You can use here also the update function of the dictionary.

File: skactiveml/tests/template_estimator.py [hunk 9]
```diff
@@ -0,0 +1,740 @@
...
+            self.estimator_class.predict_target_distribution,
+            "predict_target_distribution",
+            not_test,
+        )
+        super().test_param_test_availability()
+
+    def test_predict_target_distribution_param_X(self):
+        test_cases = []
+        X = np.array([[0, 1], [1, 0], [2, 3]])
+        test_cases += [(X, None), ("Test", ValueError)]
```
[the remark this answers is not in the document]
Comment (tpham93) [reply skactiveml/tests/template_estimator.py:683]: @AlexanderBenz can you check this again?

**Review by mherde — CHANGES_REQUESTED** [review]

**Review by tpham93 — APPROVED** [review]

**Review by mherde — APPROVED** [review]

**Pull request merged**